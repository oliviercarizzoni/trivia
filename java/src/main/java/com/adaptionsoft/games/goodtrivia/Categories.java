package com.adaptionsoft.games.goodtrivia;

import java.util.HashMap;
import java.util.Map;

class Categories {

    private final Map<Category, Question> categories;

    Categories() {
        categories = new HashMap<>();
        initQuestions();
    }

    private void initQuestions() {
        for (Category type : Category.values()) {
            categories.put(type, new Question(type));
        }
    }

    Category currentCategory(int playerPlace) {
        for (int i = 0; i <= 8; i += 4) if (playerPlace == i) return Category.POP;
        for (int i = 1; i <= 9; i += 4) if (playerPlace == i) return Category.SCIENCE;
        for (int i = 2; i <= 10; i += 4) if (playerPlace == i) return Category.SPORT;
        return Category.ROCK;
    }

    void askQuestion(int playerPlace) {
        System.out.println("The category is " + currentCategory(playerPlace).getLibelle());
        categories.get(currentCategory(playerPlace)).askQuestion();
    }
}
