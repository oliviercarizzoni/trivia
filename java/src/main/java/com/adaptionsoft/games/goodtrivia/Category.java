package com.adaptionsoft.games.goodtrivia;

enum Category {

    POP("Pop"),
    SCIENCE("Science"),
    SPORT("Sports"),
    ROCK("Rock");

    private final String libelle;

    Category(String libelle) {
        this.libelle = libelle;
    }

    public String getLibelle() {
        return libelle;
    }
}
