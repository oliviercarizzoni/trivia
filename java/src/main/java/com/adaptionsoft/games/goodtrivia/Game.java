package com.adaptionsoft.games.goodtrivia;

public class Game {

    public static final int MAX_PLAYER_PLACE = 12;
    public static final int PURSES_TO_WIN = 6;
    public static final int MIN_NB_PLAYER = 2;

    private final Players players = new Players();
    private final Categories categories = new Categories();
    private Player currentPlayer = null;

    public boolean isPlayable() {
        return (players.getNbPlayers() >= MIN_NB_PLAYER);
    }

    public void add(String playerName) {
        players.add(playerName);
        System.out.println(playerName + " was added");
        System.out.println("They are player number " + players.getNbPlayers());
    }

    public void roll(int roll) {
        checkConditionsToPlay();

        currentPlayer = players.getCurrentAndUpdate();
        System.out.println(currentPlayer + " is the current player");
        System.out.println("They have rolled a " + roll);

        if (currentPlayer.gotPenalty()) {
            if (conditionToGetRideOfPenalty(roll)) {
                currentPlayer.setPenalty(false);
                System.out.println(
                        currentPlayer + " is getting out of the penalty box");
            } else {
                System.out.println(
                        currentPlayer + " is not getting out of the penalty box");
            }
        }
        updatePlaceAndAskQuestion(roll);
    }

    private void checkConditionsToPlay() {
        if (!isPlayable()) throw new RuntimeException(
                "Game is not playable for players less than " + MIN_NB_PLAYER);
    }

    private boolean conditionToGetRideOfPenalty(int roll) {
        return roll % 2 != 0;
    }

    private void updatePlaceAndAskQuestion(int roll) {
        currentPlayer.increasePlace(roll);
        updatePlaceIfOverLimit();
        printPlayerLocation();
        categories.askQuestion(currentPlayer.getPlace());
    }

    private void updatePlaceIfOverLimit() {
        if (currentPlayer.getPlace() >= MAX_PLAYER_PLACE)
            currentPlayer.decreasePlace(MAX_PLAYER_PLACE);
    }

    private void printPlayerLocation() {
        System.out.println(currentPlayer
                           + "'s new location is "
                           + currentPlayer.getPlace());
    }

    public boolean wasCorrectlyAnswered() {
        if (currentPlayer.gotPenalty()) return true;
        System.out.println("Answer was correct!!!!");
        currentPlayer.increasePurse();
        System.out.println(currentPlayer
                           + " now has "
                           + currentPlayer.getPurse()
                           + " Gold Coins.");
        return didPlayerWin();
    }

    public boolean wrongAnswer() {
        System.out.println("Question was incorrectly answered");
        System.out.println(currentPlayer + " was sent to the penalty box");
        currentPlayer.setPenalty(true);
        return true;
    }

    private boolean didPlayerWin() {
        return !(currentPlayer.getPurse() == PURSES_TO_WIN);
    }
}
