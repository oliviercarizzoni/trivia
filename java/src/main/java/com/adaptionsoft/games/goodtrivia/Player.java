package com.adaptionsoft.games.goodtrivia;

class Player {

    private String name;
    private int place = 0;
    private int purse = 0;
    private boolean penalty = false;

    Player(String name) {
        this.name = name;
    }

    boolean gotPenalty() {
        return penalty == true;
    }

    void setPenalty(boolean penalty) {
        this.penalty = penalty;
    }

    int getPurse() {
        return purse;
    }

    void increasePurse() {
        this.purse++;
    }

    int getPlace() {
        return place;
    }

    void increasePlace(int place) {
        this.place += place;
    }

    void decreasePlace(int place) {
        this.place -= place;
    }

    @Override
    public String toString() {
        return name;
    }
}
