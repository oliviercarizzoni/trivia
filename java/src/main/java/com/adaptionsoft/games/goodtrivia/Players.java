package com.adaptionsoft.games.goodtrivia;

import java.util.ArrayList;

class Players {

    private final ArrayList<Player> players = new ArrayList<>();
    private int currentPlayer = 0;

    void add(String playerName) {
        players.add(new Player(playerName));
    }

    int getNbPlayers() {
        return players.size();
    }

    Player getCurrentAndUpdate() {
        if (isLastPlayer()) {
            int actual = currentPlayer;
            currentPlayer = 0;
            return players.get(actual);
        }
        return players.get(currentPlayer++);
    }

    private boolean isLastPlayer() {
        return currentPlayer == players.size() - 1;
    }
}
