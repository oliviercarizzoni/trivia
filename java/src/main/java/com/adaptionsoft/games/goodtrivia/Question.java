package com.adaptionsoft.games.goodtrivia;

import java.util.LinkedList;
import java.util.stream.IntStream;

public class Question {

    private static final int NB_QUESTIONS = 50;

    private final LinkedList<String> qlist = new LinkedList<>();
    private final Category type;

    public Question(Category type) {
        this.type = type;
        initQuestions();
    }

    void askQuestion() {
        System.out.println(qlist.removeFirst());
    }

    private void initQuestions() {
        IntStream.range(0, NB_QUESTIONS)
                 .forEachOrdered(i -> qlist.addLast(type.getLibelle() + " Question " + i));
    }
}
