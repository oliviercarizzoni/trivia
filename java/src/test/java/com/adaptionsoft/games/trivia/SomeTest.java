package com.adaptionsoft.games.trivia;

import com.adaptionsoft.games.goodtrivia.Game;
import org.junit.Test;

import java.util.stream.IntStream;

import static org.approvaltests.Approvals.verify;

public class SomeTest extends RecordTestCase {

    @Test
    public void test_standard_game() {
        recordAndTestGame(3, 5, 3);
    }

    @Test
    public void test_5players_10roll_0fails() {
        recordAndTestGame(5, 10, 0);
    }

    @Test(expected = RuntimeException.class)
    public void test_1player_26roll_0fails() {
        recordAndTestGame(1, 26, 0);
    }

    @Test
    public void can_play_with_more_than_5_players() {
        recordAndTestGame(6, 1, 1);
    }

    private void recordAndTestGame(int nbPlayers, int rollNumber, int nbFails) {
        recording();
        runGame(nbPlayers, rollNumber, nbFails);
        stopRecording();
        verify(getRecording());
    }

    private void runGame(int nbPlayers, int rollNumber, int nbFails) {
        Game aGame = new Game();
        boolean notAWinner;
        IntStream.rangeClosed(1, nbPlayers).forEach(nb -> aGame.add("User" + nb));
        do {
            aGame.roll(rollNumber);
            if (nbFails > 0) {
                notAWinner = aGame.wrongAnswer();
                nbFails--;
            } else {
                notAWinner = aGame.wasCorrectlyAnswered();
            }
        } while (notAWinner);
    }

}
