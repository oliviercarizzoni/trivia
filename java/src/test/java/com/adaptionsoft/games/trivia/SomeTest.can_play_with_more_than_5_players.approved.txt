User1 was added
They are player number 1
User2 was added
They are player number 2
User3 was added
They are player number 3
User4 was added
They are player number 4
User5 was added
They are player number 5
User6 was added
They are player number 6
User1 is the current player
They have rolled a 1
User1's new location is 1
The category is Science
Science Question 0
Question was incorrectly answered
User1 was sent to the penalty box
User2 is the current player
They have rolled a 1
User2's new location is 1
The category is Science
Science Question 1
Answer was correct!!!!
User2 now has 1 Gold Coins.
User3 is the current player
They have rolled a 1
User3's new location is 1
The category is Science
Science Question 2
Answer was correct!!!!
User3 now has 1 Gold Coins.
User4 is the current player
They have rolled a 1
User4's new location is 1
The category is Science
Science Question 3
Answer was correct!!!!
User4 now has 1 Gold Coins.
User5 is the current player
They have rolled a 1
User5's new location is 1
The category is Science
Science Question 4
Answer was correct!!!!
User5 now has 1 Gold Coins.
User6 is the current player
They have rolled a 1
User6's new location is 1
The category is Science
Science Question 5
Answer was correct!!!!
User6 now has 1 Gold Coins.
User1 is the current player
They have rolled a 1
User1 is getting out of the penalty box
User1's new location is 2
The category is Sports
Sports Question 0
Answer was correct!!!!
User1 now has 1 Gold Coins.
User2 is the current player
They have rolled a 1
User2's new location is 2
The category is Sports
Sports Question 1
Answer was correct!!!!
User2 now has 2 Gold Coins.
User3 is the current player
They have rolled a 1
User3's new location is 2
The category is Sports
Sports Question 2
Answer was correct!!!!
User3 now has 2 Gold Coins.
User4 is the current player
They have rolled a 1
User4's new location is 2
The category is Sports
Sports Question 3
Answer was correct!!!!
User4 now has 2 Gold Coins.
User5 is the current player
They have rolled a 1
User5's new location is 2
The category is Sports
Sports Question 4
Answer was correct!!!!
User5 now has 2 Gold Coins.
User6 is the current player
They have rolled a 1
User6's new location is 2
The category is Sports
Sports Question 5
Answer was correct!!!!
User6 now has 2 Gold Coins.
User1 is the current player
They have rolled a 1
User1's new location is 3
The category is Rock
Rock Question 0
Answer was correct!!!!
User1 now has 2 Gold Coins.
User2 is the current player
They have rolled a 1
User2's new location is 3
The category is Rock
Rock Question 1
Answer was correct!!!!
User2 now has 3 Gold Coins.
User3 is the current player
They have rolled a 1
User3's new location is 3
The category is Rock
Rock Question 2
Answer was correct!!!!
User3 now has 3 Gold Coins.
User4 is the current player
They have rolled a 1
User4's new location is 3
The category is Rock
Rock Question 3
Answer was correct!!!!
User4 now has 3 Gold Coins.
User5 is the current player
They have rolled a 1
User5's new location is 3
The category is Rock
Rock Question 4
Answer was correct!!!!
User5 now has 3 Gold Coins.
User6 is the current player
They have rolled a 1
User6's new location is 3
The category is Rock
Rock Question 5
Answer was correct!!!!
User6 now has 3 Gold Coins.
User1 is the current player
They have rolled a 1
User1's new location is 4
The category is Pop
Pop Question 0
Answer was correct!!!!
User1 now has 3 Gold Coins.
User2 is the current player
They have rolled a 1
User2's new location is 4
The category is Pop
Pop Question 1
Answer was correct!!!!
User2 now has 4 Gold Coins.
User3 is the current player
They have rolled a 1
User3's new location is 4
The category is Pop
Pop Question 2
Answer was correct!!!!
User3 now has 4 Gold Coins.
User4 is the current player
They have rolled a 1
User4's new location is 4
The category is Pop
Pop Question 3
Answer was correct!!!!
User4 now has 4 Gold Coins.
User5 is the current player
They have rolled a 1
User5's new location is 4
The category is Pop
Pop Question 4
Answer was correct!!!!
User5 now has 4 Gold Coins.
User6 is the current player
They have rolled a 1
User6's new location is 4
The category is Pop
Pop Question 5
Answer was correct!!!!
User6 now has 4 Gold Coins.
User1 is the current player
They have rolled a 1
User1's new location is 5
The category is Science
Science Question 6
Answer was correct!!!!
User1 now has 4 Gold Coins.
User2 is the current player
They have rolled a 1
User2's new location is 5
The category is Science
Science Question 7
Answer was correct!!!!
User2 now has 5 Gold Coins.
User3 is the current player
They have rolled a 1
User3's new location is 5
The category is Science
Science Question 8
Answer was correct!!!!
User3 now has 5 Gold Coins.
User4 is the current player
They have rolled a 1
User4's new location is 5
The category is Science
Science Question 9
Answer was correct!!!!
User4 now has 5 Gold Coins.
User5 is the current player
They have rolled a 1
User5's new location is 5
The category is Science
Science Question 10
Answer was correct!!!!
User5 now has 5 Gold Coins.
User6 is the current player
They have rolled a 1
User6's new location is 5
The category is Science
Science Question 11
Answer was correct!!!!
User6 now has 5 Gold Coins.
User1 is the current player
They have rolled a 1
User1's new location is 6
The category is Sports
Sports Question 6
Answer was correct!!!!
User1 now has 5 Gold Coins.
User2 is the current player
They have rolled a 1
User2's new location is 6
The category is Sports
Sports Question 7
Answer was correct!!!!
User2 now has 6 Gold Coins.
